crates/holochain_integrity_types/src/capability/grant.rs

```diff
@@ -87,37 +87,43 @@ impl CapGrant {
     /// if a grant is valid in a standalone way.
     pub fn is_valid(
         &self,
-        check_function: &GrantedFunction,
-        check_agent: &AgentPubKey,
-        check_secret: Option<&CapSecret>,
+        given_function: &GrantedFunction,
+        given_agent: &AgentPubKey,
+        given_secret: Option<&CapSecret>,
     ) -> bool {
         match self {
             // Grant is always valid if the author matches the check agent.
-            CapGrant::ChainAuthor(author) => author == check_agent,
+            CapGrant::ChainAuthor(author) => author == given_agent,
             // Otherwise we need to do more work…
             CapGrant::RemoteAgent(ZomeCallCapGrant {
                 access, functions, ..
             }) => {
                 // The checked function needs to be in the grant…
                 let granted = match functions {
                     GrantedFunctions::All => true,
-                    GrantedFunctions::Listed(fns) => fns.contains(check_function),
+                    GrantedFunctions::Listed(fns) => fns.contains(given_function),
                 };
                 granted
                 // The agent needs to be valid…
                 && match access {
                     // The grant is assigned so the agent needs to match…
-                    CapAccess::Assigned { assignees, .. } => assignees.contains(check_agent),
+                    CapAccess::Assigned { assignees, .. } => assignees.contains(given_agent),
                     // The grant has no assignees so is always valid…
                     _ => true,
                 }
-                // The secret needs to match…
+                // The secret needs to match, if provided...
                 && match access {
-                    // Unless the extern is unrestricted.
+                    // ...unless the grant requires no secret
                     CapAccess::Unrestricted => true,
+                    CapAccess::Assigned { secret: None, .. } => true,
+
                     // note the PartialEq implementation is constant time for secrets
-                    CapAccess::Transferable { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
-                    CapAccess::Assigned { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
+                    // so this is safe from timing attacks
+
+                    // Transferable grants always require a matching secret
+                    CapAccess::Transferable { secret, .. } => Some(secret) == given_secret,
+                    // If a secret is required for the Assigned grant, a matching secret must be provided
+                    CapAccess::Assigned { secret: Some(secret), .. } => given_secret.map(|given| secret == given).unwrap_or(false)
```

Comment:
I didn't do that because that would have the effect of penalizing someone for providing a secret for a cap that doesn't require one. I wanted it so that if the Assigned cap grant has a None secret, it's valid to provide a secret or not.

Reply:
oh ha ha, you're right!